app/src/marketplace/containers/EditProductPage2/ProductName.jsx

```diff
@@ -0,0 +1,37 @@
+// @flow
+
+import React from 'react'
+
+import useProduct from '../ProductController/useProduct'
+import useValidation from '../ProductController/useValidation'
+import useProductUpdater from '../ProductController/useProductUpdater'
+
+const ProductName = () => {
+    const product = useProduct()
+    const { updateProduct } = useProductUpdater()
+    const { status } = useValidation('name')
+
+    const onChange = (name) => {
+        updateProduct('Update name', (p) => ({
+            ...p,
+            name,
+        }))
+    }
```

Comment:
So after working with the editor code for a while, now we have hooks I've been considering consolidating all undo stack actions into something like `useProductActions.js` then doing something like:

```js
import { useProductActions } from '../../ProductController/useProductActions' 
…
const { updateName } = useProductActions()
```

What do you think?

Downside is that related code is no longer co-located.

Reply:
Ok, that would sound pretty good - would the undo function and the product data come also from there?
```
const { undo, product, updateName } = useProductActions()
// set name
updateName('my product')
console.log(product)
// undo previous change
undo()
```
Would it make sense for the validation to exist in the same place?

Reply:
Maybe the undo/redo but I'd probably keep the product separate so as to avoid unnecessary rerenders, semi-related: https://twitter.com/jevakallio/status/1153371850250117121

Reply:
Ok, updated to include product actions hook with a couple updater functions. @timoxley Is that in line what you were thinking?